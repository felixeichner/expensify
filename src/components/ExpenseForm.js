import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor (props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? numeral(props.expense.amount / 100).format('0.00') : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\,\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide a description and a valid amount' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(numeral(this.state.amount).value(), 10) * 100,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf()
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p className="error-message">{this.state.error}</p>}
        <form id="expense-form" onSubmit={this.onSubmit}>
          <table>
            <tbody>
              <tr>
                <th>Description</th>
                <td>
                  <input 
                    type="text"
                    value={this.state.description}
                    placeholder="Description"
                    autoFocus
                    onChange={this.onDescriptionChange}
                  />
                </td>
              </tr>
              <tr>
                <th>Amount (â‚¬)</th>
                <td>
                  <input
                    type="text"
                    value={this.state.amount}
                    placeholder="Amount"
                    onChange={this.onAmountChange}
                  />
                </td>
              </tr>
              <tr>
                <th>Note</th>
                <td>
                  <textarea
                    value={this.state.note}
                    placeholder="Add a note for your expense (optional)"
                    onChange={this.onNoteChange}
                  >
                  </textarea>
                </td>
              </tr>
              <tr>
                <th>Date</th>
                <td>
                  <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    displayFormat={"DD. MMM YYYY"}
                  />
                </td>
              </tr>
            </tbody>
          </table>
          <p style={{marginLeft:'30px'}}>
            <button type="submit">{this.props.expense ? 'Update Expense' : 'Add Expense'}</button>
          </p>
        </form>
      </div>
    )
  }
};