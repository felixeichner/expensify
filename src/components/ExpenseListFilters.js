import React from 'react';
import { connect } from 'react-redux';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  render() {
    return (
      <div id={'filter-bar'}>
        <span>
          <strong>Search for: </strong>
          <input
            type="text"
            value={this.props.filters.text}
            onChange={(e) => {
              this.props.dispatch(setTextFilter(e.target.value));
            }}
          />
        </span>
        <span>
          <strong>Sort by: </strong>
          <select
            value={this.props.filters.sortBy}
            onChange={(e) => {
              this.props.dispatch((e.target.value === 'amount') ? sortByAmount() : sortByDate());
            }}
          >
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>
        </span>
        <span>
          <strong>Filter by date: </strong>
          <DateRangePicker
            startDate={this.props.filters.startDate}
            startDateId="startDateId"
            endDate={this.props.filters.endDate}
            endDateId="endDateId"
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            showClearDates={true}
            numberOfMonths={1}
            isOutsideRange={() => false}
            displayFormat={"DD. MMM YYYY"}
          />
        </span>
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
};

export default connect(mapStateToProps)(ExpenseListFilters);