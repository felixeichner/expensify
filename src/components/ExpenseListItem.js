import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({ id, description, note, amount, createdAt }) => {
  return (
    <tr>
      <td><Link to={`/edit/${id}`}><strong>{description}</strong></Link></td>
      <td>{note}</td>
      <td>â‚¬ {(amount / 100).toFixed(2)}</td>
      <td>{moment(createdAt).format("DD. MMM YYYY")}</td>
    </tr>
  )
};

export default ExpenseListItem;