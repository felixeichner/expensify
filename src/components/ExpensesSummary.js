import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import 'numeral/locales/de';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

numeral.locale('de');

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  return (
    <div className="page-header">
      Viewing <span className="bold">{expenseCount}</span>
      {expenseCount === 1 ? ' expense ' : ' expenses '}
      totalling <span className="bold">{numeral(expensesTotal / 100).format('$ 0,0.00')}</span>
    </div>
  )
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: getExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);