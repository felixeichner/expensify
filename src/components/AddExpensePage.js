import React from 'react';
import ExpenseForm from './ExpenseForm';

const AddExpense = () => (
  <div>
    <h4>Add Expense:</h4>
    <ExpenseForm />
  </div>
);

export default AddExpense;