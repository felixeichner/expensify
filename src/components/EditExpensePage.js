import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

const EditExpense = (props) => {
  return (
    <div>
      <h4>Edit Expense</h4>
      <button
        onClick={() => {
          props.dispatch(removeExpense({ id: props.expense.id }));
          props.history.push('/');
        }}
        className='button-alert'
      >
        Remove
      </button>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          props.dispatch(editExpense(props.expense.id, expense));
          props.history.push('/');
        }}
      />
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
};

export default connect(mapStateToProps)(EditExpense);