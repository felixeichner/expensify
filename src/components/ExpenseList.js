import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = (props) => {
  return (
    <div className="expenses-list">
      <div className="expenses-list__header">
        <span>Expenses</span>
        <span>Amount</span>
      </div>
      {
        props.expenses.length === 0 ? (
          <div className="expenses-list__content bold">No expenses to list!</div>
        ) : (
          props.expenses.map((expense) => (
            <ExpenseListItem key={expense.id} {...expense} />
          ))
        )
      }
    </div>
  )
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
};

export default connect(mapStateToProps)(ExpenseList);