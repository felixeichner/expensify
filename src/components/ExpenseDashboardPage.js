import React from 'react';
import { Link } from 'react-router-dom';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboard = () => (
  <div className="content-container">
    <ExpensesSummary />
    <Link to="/create">
      <button className="button">Add Expense</button>
    </Link>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;