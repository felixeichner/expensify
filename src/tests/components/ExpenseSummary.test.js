import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render ExpensesSummary with no expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={0} expensesTotal={0} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with one expense', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={12354} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={625437} />);
  expect(wrapper).toMatchSnapshot();
});