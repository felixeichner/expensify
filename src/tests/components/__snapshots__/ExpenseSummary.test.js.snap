// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with multiple expenses 1`] = `
<div
  className="page-header"
>
  Viewing 
  <span
    className="bold"
  >
    23
  </span>
   expenses 
  totalling 
  <span
    className="bold"
  >
    € 6 254,37
  </span>
</div>
`;

exports[`should render ExpensesSummary with no expenses 1`] = `
<div
  className="page-header"
>
  Viewing 
  <span
    className="bold"
  >
    0
  </span>
   expenses 
  totalling 
  <span
    className="bold"
  >
    € 0,00
  </span>
</div>
`;

exports[`should render ExpensesSummary with one expense 1`] = `
<div
  className="page-header"
>
  Viewing 
  <span
    className="bold"
  >
    1
  </span>
   expense 
  totalling 
  <span
    className="bold"
  >
    € 123,54
  </span>
</div>
`;
